scripts/generate_hamilton_input_UPL.py

```diff
@@ -0,0 +1,100 @@
+            if input.type == 'Analyte':
+                output = self.process.outputs_per_input(input.id, Analyte=True)
+                #the script is only compatible with 1 output for each input i.e. replicates are not allowed
+                if len(output) >1:
+                    print('Multiple outputs found for an input %s. This step is not compatible with replicates.' %(input.name))
+                    sys.exit(1)
+
+
+                # build a list of the unique input containers for checking that no more than 9 are present due to
+                #deck limit on Hamilton and for sorting the sample locations by input plate. Build a list of unique
+                #output containers as no more than 1 plate
+                unique_input_containers.add(input.container.name)
+
+                unique_output_containers.add(output[0].container.name)
+
+                #assemble each line of the Hamilton input file in the correct structure for the Hamilton
+                csv_line = [input.container.name, input.location[1], output[0].container.name, output[0].location[1],
+                            self.process.udf['DNA Volume (uL)'], '0']
+                #build a dictionary of the lines for the Hamilton input file with a key that facilitates the lines being
+                #by input container then column then row
+                csv_dict[input.container.name+input.location[1]] = csv_line
+
+        #check the number of input containers
+        if len(unique_input_containers) > 9:
+            print('Maximum number of input plates is 9. There are %s output plates in the step.' % (str(len(unique_input_containers))))
+            sys.exit(1)
+        #check the number of output containers
+        if len(unique_output_containers) > 1:
+            print('Maximum number of output plates is 1. There are %s output plates in the step.' % (str(len(unique_output_containers))))
+            sys.exit(1)
+        #define the rows and columns in the input plate (standard 96 well plate pattern)
+        rows = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H']
+        columns = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']
+
+        #add the lines to the csv_array that will be used to write the Hamilton input file
+        for unique_input_container in sorted(unique_input_containers):
+            for column in columns:
+                for row in rows:
+                    if unique_input_container + row + ":" + column in csv_dict.keys():
+                        csv_array.append(csv_dict[unique_input_container + row + ":" + column])
+
+
+        #create and write the Hamilton input file, this must have the hamilton_input argument as the prefix as this is used by
+        #Clarity LIMS to recognise the file and attach it to the step
+        with open(self.hamilton_input + '-hamilton_input.csv', 'w') as csvFile:
+            writer = csv.writer(csvFile)
+            writer.writerows(csv_array)
+        csvFile.close()
```

Comment:
We don't need to close the file after, because the `with` context manager already does this